### ISSUE TITLE
Function not implemented remote: fatal: failed to write object

### ISSUE THREAD

**Issue by gregstephensen** [git-for-windows/git#2764]
- [ ] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

git version 2.27.0.windows.1
cpu: x86_64
built from commit: 907ab1011dce9112700498e034b974ba60f8b407
sizeof-long: 4
sizeof-size_t: 8

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.17763.1282]
(c) 2018 Microsoft Corporation. All rights reserved.


 - What options did you set as part of the installation? Or did you choose the
   defaults?

Editor Option: VIM
Custom Editor Path:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

I'm running a VPN client to connect to work network, when I completely disconnect and reconnect I can push to the bare respos on the shared drive. 

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

BASH

 - What commands did you run to trigger this issue? 

Git Push


 - What did you expect to occur after running these commands?

For it to push my changes

 - What actually happened instead?

Enumerating objects: 15, done.
Counting objects: 100% (15/15), done.
Delta compression using up to 4 threads
Compressing objects: 100% (8/8), done.
Writing objects: 100% (8/8), 906 bytes | 453.00 KiB/s, done.
Total 8 (delta 7), reused 0 (delta 0), pack-reused 0
remote: error: unable to write file //test/test$/test Repos/./objects/incoming-a14396/ab/002244bbe766a1132131724192cac35b58ccdb: Function not implemented
remote: fatal: failed to write object
error: remote unpack failed: unpack-objects abnormal exit
To //test/test$/test Repos

**Comment by gregstephensen** [comment]
Although it didn't solve my problem, by moving my repos from the shared network drive, to a faster windows server seems to have allowed me to continue.

**Comment by dscho** [comment]
> //test/test$/test Repos/./objects/incoming-a14396/ab/002244bbe766a1132131724192cac35b58ccdb: Function not implemented

This looks like a network share. There are quite a few matches when you search this issue tracker for "Function not implemented".

I guess that your scenario might be similar/identical to what has been described here: https://github.com/git-for-windows/git/issues/2710#issuecomment-666244835

**Comment by gregstephensen** [comment]
Yeah I've looked them up, but the problem is that it works sometimes, but then suddenly it doesn't and I have to disconnect from the VPN and reconnect. I can still look up files in the file share like normal, it's only Git not working.

**Comment by dscho** [comment]
Can you share details about the server side?

And can you [rebuild Git with patches](https://github.com/git-for-windows/git/wiki/Building-Git)? If so, I would suggest to instrument the code to learn more.

Let's investigate where the error message comes from in this case:

```
$ git grep "unable to write file" \*.c
apply.c:        return error_errno(_("unable to write file '%s' mode %o"),
entry.c:                        return error("unable to write file %s", path);
sha1-file.c:                    return error_errno(_("unable to write file %s"), filename);
```

Since the error message is followed by the system error message "Function not implemented", it will be the one in `sha1-file.c`. Here is the code: https://github.com/git-for-windows/git/blob/77982caf269b7ee713a76da2bcf260c34d3bf7a7/sha1-file.c#L1757

The bigger context looks like this: https://github.com/git-for-windows/git/blob/77982caf269b7ee713a76da2bcf260c34d3bf7a7/sha1-file.c#L1737-L1760

Looking at the [blame output](https://github.com/git-for-windows/git/blame/v2.28.0.windows.1/sha1-file.c#L1737-L1760), it would appear that this "Coda hack" is over 15 years old already.

The problem might be with one of the `rename()`, `unlink()` or any other "syscall" that is emulated in `compat/mingw.c` and that tries to map Windows error codes to appropriate `errno` values (probably not always succeeding).

I could imagine that the most useful approach might be to instrument [`err_win_to_posix()`](https://github.com/git-for-windows/git/blob/77982caf269b7ee713a76da2bcf260c34d3bf7a7/compat/mingw.c#L32-L149) such that it prints the mapping just before returning. Something like

```c
if (getenv("MY_DEBUG_ERRNO_VALUES"))
    fprintf(stderr, "winerr %ld mapped to %d (%s)\n", err, errno, strerror(errno));
```

and then running the newly-built Git with `MY_DEBUG_ERRNO_VALUES=1 ./bin-wrappers/git -C /path/to/local/repository [...]`.

**Comment by gregstephensen** [comment]
I used git version 2.18 and it seems to work fine now.

**Comment by gregstephensen** [comment]
Scratch that, nothing has worked.

**Comment by dscho** [comment]
> And can you [rebuild Git with patches](https://github.com/git-for-windows/git/wiki/Building-Git)? If so, I would suggest to instrument the code to learn more.

@gregstephensen I'd be still interested in pursuing this approach if you're still invested in the issue. I fear that we don't have many alternatives left to investigate further.

**Comment by gregstephensen** [comment]
I downloaded the SDK, what file do I need to edit cause I can't find it in the SDK.

**Comment by dscho** [comment]
> I downloaded the SDK, what file do I need to edit cause I can't find it in the SDK.

It's in the Git source code. The easiest way to get that is `sdk cd git`. You can then build with `sdk build`, or `make -j$(nproc)`.

**Comment by dscho** [comment]
Closing as stale.
